File: Projekti/C#-Native-Benchmark/Repository/UserRepository.cs
using MySql.Data.MySqlClient;
using Native.Dto;
using Native.Model;
using Native.Repository.Interface;
using Native.Settings;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Native.Repository
{
    public class UserRepository : IUserRepository
    {

        private readonly string _connectionString;

        public UserRepository(IDatabaseConnection connection)
        {
            _connectionString = connection.ConnectionString;
        }

        public List<User> GetUsersBornAfter(DateTime year)
        {
            List<User> users = new List<User>();
            using MySqlConnection connection = new(_connectionString);
            connection.Open();
            MySqlCommand command = connection.CreateCommand();

            command.CommandText = "SELECT * FROM Users WHERE date_of_birth >= @date";
            command.Parameters.AddWithValue("@date", year);

            using MySqlDataReader reader = command.ExecuteReader();

            while (reader.Read())
            {
                users.Add(new User(reader.GetInt32(0), reader.GetString(1), reader.GetString(2), reader.GetString(3), reader.GetDateTime(4)));
            }
       
            return users;
        }

        public List<UserPostCount> GetNumberOfPostsForEachUser()
        {
            List<UserPostCount> users = new List<UserPostCount>();

            using MySqlConnection connection = new(_connectionString);
            connection.Open();
            MySqlCommand command = connection.CreateCommand();

            command.CommandText = "SELECT u.user_id, u.email, COUNT(p.post_id) " +
                                  "FROM Users AS u " +
                                  "LEFT OUTER JOIN Posts AS p ON u.user_id = p.user_id " +
                                  "GROUP BY u.user_id";

            using MySqlDataReader reader = command.ExecuteReader();

            while (reader.Read()) 
            {
                users.Add(new UserPostCount(reader.GetInt32(0), reader.GetString(1), reader.GetInt32(2)));
            }

            return users;
        }

        public List<UserPostCount> GetTopFiveUsersByPostCount()
        {
            List<UserPostCount> users = new List<UserPostCount>();

            using MySqlConnection connection = new(_connectionString);
            connection.Open();
            MySqlCommand command = connection.CreateCommand();

            command.CommandText = "SELECT u.user_id, u.email, COUNT(p.post_id) as total_posts " +
                "FROM Users as u, Posts as p " +
                "WHERE u.user_id = p.user_id " +
                "GROUP BY u.user_id " +
                "ORDER BY total_posts DESC " +
                "LIMIT 5";

            using MySqlDataReader reader = command.ExecuteReader();

            while (reader.Read())
            {
                users.Add(new UserPostCount(reader.GetInt32(0), reader.GetString(1), reader.GetInt32(2)));
            }

            return users;
        }

        public List<UserProfile> GetUsersWithKeywordInBio(string keyword)
        {
            List<UserProfile> users = new List<UserProfile>();

            using MySqlConnection connection = new(_connectionString);
            connection.Open();
            MySqlCommand command = connection.CreateCommand();

            command.CommandText = "SELECT u.user_id, u.email, p.bio " +
                "FROM Users u, Profiles p " +
                "WHERE u.user_id = p.user_id " +
                "AND p.bio LIKE @param";

            command.Parameters.AddWithValue("@param", "%" + keyword + "%");

            using MySqlDataReader reader = command.ExecuteReader();

            while (reader.Read())
            {
                users.Add(new UserProfile(reader.GetInt32(0), reader.GetString(1), reader.GetString(2)));
            }

            return users;
        }

        public void InsertUser(User user)
        {
            using MySqlConnection connection = new(_connectionString);
            connection.Open();
            MySqlCommand command = connection.CreateCommand();

            MySqlTransaction transaction = connection.BeginTransaction();

            command.CommandText = "INSERT INTO Users VALUES (@user_id, @first_name, @last_name, @email, @date_of_birth)";
            command.Parameters.AddWithValue("@first_name", user.FirstName);
            command.Parameters.AddWithValue("@last_name", user.LastName);
            command.Parameters.AddWithValue("@email", user.Email);
            command.Parameters.AddWithValue("@date_of_birth", user.DateOfBirth);
            command.Parameters.AddWithValue("@user_id", user.UserId);

            command.ExecuteNonQuery();
            transaction.Rollback();
        }
    }
}

File: Projekti/Python-Native-Benchmark/repository/user_repository.py
from model.user import User
from mysql.connector.connection import MySQLConnection

class UserRepository:
    def __init__(self, conn : MySQLConnection):
        self.conn = conn

    def fetch_all_users(self):
        with self.conn.cursor() as cursor:
            cursor.execute("SELECT * FROM users")
            return cursor.fetchall()

    def fetch_user_by_id(self, user_id):
        with self.conn.cursor() as cursor:
            cursor.execute("SELECT * FROM users WHERE id = %s", (user_id,))
            return cursor.fetchone()

    def add_user(self, first_name, last_name, email, date):
        with self.conn.cursor() as cursor:
            cursor.execute(
                "INSERT INTO Users (first_name, last_name, email, date_of_birth) VALUES (%s, %s, %s, %s)",
                (first_name, last_name, email, date)
            )
            self.conn.commit()
            return cursor.lastrowid 
        
    def get_users_born_after(self, year):

        query = "SELECT * FROM Users WHERE date_of_birth >= (%s)"
        users = [];
        params = [];
        params.append(year)

        with self.conn.cursor(buffered=True) as cursor:
            cursor.execute(query, params)

        for (user_id, first_name, last_name, email, date_of_birth) in cursor:
            users.append(User(user_id, first_name, last_name, email, date_of_birth))

        return users

    def get_number_of_posts_for_each_user(self):

        query =  "SELECT u.user_id, u.email, COUNT(p.post_id) "
        query += "FROM Users AS u "
        query += "LEFT OUTER JOIN Posts AS p ON u.user_id = p.user_id "
        query += "GROUP BY u.user_id"

        users = []

        with self.conn.cursor(buffered=True) as cursor:
            cursor.execute(query)

        for (user_id, email, count) in cursor:
            users.append(tuple((user_id, email, count)))

        return users
    
    def get_top_five_users(self):

        query =  "SELECT u.user_id, u.email, COUNT(p.post_id) " 
        query += "FROM Users AS u " 
        query += "LEFT OUTER JOIN Posts AS p ON u.user_id = p.user_id " 
        query +=  "GROUP BY u.user_id"

        users = []

        with self.conn.cursor(buffered=True) as cursor: 
            cursor.execute(query)

        for (user_id, email, count) in cursor:
            users.append(tuple((user_id, email, count)))

        return users
    
    def get_users_by_keyword(self, keyword):

        query =  "SELECT u.user_id, u.email, p.bio "
        query += "FROM Users u, Profiles p "
        query += "WHERE u.user_id = p.user_id "
        query += "AND p.bio LIKE (%s)"

        users = [] 
        params = []
        params.append(keyword + "%")

        with self.conn.cursor(buffered=True) as cursor:
            cursor.execute(query, params)

        for (user_id, email, bio) in cursor: 
            users.append(tuple((user_id, email, bio)))

        return users

        
